### PR TITLE
BBS-142 port ting entity to SAL

### PR DESCRIPTION
**Pull request by danquah**

### CODE REVIEW

File: modules/ding_loan/plugins/content_types/loans.inc [hunk 1]
```diff
@@ -179,9 +179,10 @@ function ding_loan_loans_form($form, &$form_state, $account, $loans = array()) {
         // Patron have this item in loans list, but opensearch can't find this
         // record for current agencyId, so we load it separately.
         // Here we are loading TingClientObject.
-        // TODO BBS-SAL: should use a direct provider invocation instead and
-        // ting_get_object_sal should be dropped.
-        $ting_object = ting_get_object_sal($loan->ding_entity_id);
+        $ting_object = ding_provider_invoke('search', 'object_load', [$loan->ding_entity_id]);
```
Comment (kasperg) [inline modules/ding_loan/plugins/content_types/loans.inc]: Code comment mentions OpenSearch. How about updating it?

File: modules/ding_loan/plugins/content_types/loans.inc [hunk 2]
```diff
@@ -179,9 +179,10 @@ function ding_loan_loans_form($form, &$form_state, $account, $loans = array()) {
...
+        if (!empty($ting_object) && is_array($ting_object)) {
+          $ting_object = reset($ting_object);
+        }
```
Comment (kasperg) [inline modules/ding_loan/plugins/content_types/loans.inc]: Why is this needed?

It seems overly verbose if we have a contract that if a single id is passed in then a single object should also be returned.

File: modules/opensearch/src/OpenSearchTingObject.php [hunk 3]
```diff
@@ -31,26 +32,73 @@ class OpenSearchTingObject implements TingObjectInterface {
...
    */
-  protected $dingId;
+  protected $ownerId;
 
   /**
-   * @var string Id of the owner of the object, eg. the Agency.
+   * @var TingRelation[] list of materials related to this material.
    */
-  protected $ownerId;
+  protected $relations;
```
Comment (kasperg) [inline modules/opensearch/src/OpenSearchTingObject.php]: Do we need this property? `getRelations()` do not seem to use it.

File: modules/opensearch/src/OpenSearchTingObject.php [hunk 4]
```diff
@@ -31,26 +32,73 @@ class OpenSearchTingObject implements TingObjectInterface {
...
+    $this->openSearchObject->relations = array();
+    $relation_objects = array();
+    if (isset($this->openSearchObject->relationsData)) {
+      $entity_ids = array();
+      foreach ($this->openSearchObject->relationsData as $record) {
+        if (isset($record->relationUri) && isset($record->relationObject)) {
+          $entity_ids[] = $record->relationUri;
+        }
+      }
+      if ($entity_ids) {
```
Comment (kasperg) [inline modules/opensearch/src/OpenSearchTingObject.php]: Perhaps `!empty($entity_ids)` here to reduce implicit type juggling.

File: modules/opensearch/src/OpenSearchTingObject.php [hunk 5]
```diff
@@ -89,15 +137,109 @@ public function setOwnerId($owner_id) {
...
+  public function getClassification() {
+    // Get the first classification.
+    $classification = $this->firstEntry($this->getRecordEntry('dc:subject', 'dkdcplus:DK5'));
+
+    if (empty($classification)) {
+      return FALSE;
+    }
+
+    // Ignore the shorthand for "Skønlitteratur".
+    return $classification === 'sk' ? '' : $classification;
```
Comment (kasperg) [inline modules/opensearch/src/OpenSearchTingObject.php:166]: Returning empty string and false for empty results is odd but perhaps not the time or place to change here.

File: modules/opensearch/src/OpenSearchTingObject.php [hunk 6]
```diff
@@ -89,15 +137,109 @@ public function setOwnerId($owner_id) {
...
+    $url = '';
+    // Try to find the online url from relation data, which requires us to get
+    // relations. First check if relations are set; if not do another request
+    // to get relations.
+    if (!isset($this->openSearchObject->relationsData)) {
+      $ting_client_object = ting_get_object($this->getId(), FALSE, TRUE);
+      if (isset($ting_client_object->relationsData)) {
+        $this->openSearchObject->relationsData = $ting_client_object->relationsData;
+      }
+    }
```
Comment (kasperg) [inline modules/opensearch/src/OpenSearchTingObject.php]: This seems to be duplicated from `getRelations()`. Could that be avoided?

File: modules/opensearch/src/OpenSearchTingObject.php [hunk 7]
```diff
@@ -133,7 +275,30 @@ public function getFormat() {
    * {@inheritdoc}
    */
   public function isPartOf() {
-    return $this->getRecordEntry('dcterms:isPartOf');
+    $search_only_fields = ['oss:sort'];
```
Comment (kasperg) [inline modules/opensearch/src/OpenSearchTingObject.php:286]: This is used multiple times. Consider extracting this to a class variable or constant.

File: modules/opensearch/src/OpenSearchTingObject.php [hunk 8]
```diff
@@ -133,7 +275,30 @@ public function getFormat() {
...
+    return $this->recordsFlatten($ispartof);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getContributors() {
+    $search_only_fields = ['oss:sort'];
+    $contributors = $this->filterRecordsExclude($this->getRecordLevel('dc:contributor'), $search_only_fields);
+    // We may have multiple field containing contributes, flatten that multi-
```
Comment (kasperg) [inline modules/opensearch/src/OpenSearchTingObject.php]: contributes => contributors

File: modules/opensearch/src/OpenSearchTingObject.php [hunk 9]
```diff
@@ -89,15 +137,109 @@ public function setOwnerId($owner_id) {
...
+  public function getLanguage() {
+    return $this->firstEntry($this->getRecordEntry('dc:language'));
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function isOnline() {
+    // Check if the material has its own URI - if so we assume it is online.
+    return !empty($this->openSearchObject->record['dc:identifier']['dcterms:URI']);
```
Comment (kasperg) [inline modules/opensearch/src/OpenSearchTingObject.php]: Could this use one of our helper methods e.g. `getRecordEntry()`?

File: modules/opensearch/src/OpenSearchTingObject.php [hunk 10]
```diff
@@ -224,7 +410,20 @@ public function getReplacedBy() {
...
+      return $isbn;
+    }
+
+    // Get ISBN numbers.
+    $isbn = $this->openSearchObject->record['dc:identifier']['dkdcplus:ISBN'];
+    foreach ($isbn as $k => $number) {
+      $isbn[$k] = str_replace(array(' ', '-'), '', $number);
+    }
+    rsort($isbn);
+    return $isbn;
```
Comment (kasperg) [inline modules/opensearch/src/OpenSearchTingObject.php:434]: This method is changed from returning a string to an array of strings. Is that intentional?

I just noticed the change. I have not been able to extract what the calling code expects.

File: modules/ting/src/TingObject.php [hunk 11]
```diff
@@ -17,38 +17,48 @@ class TingObject implements TingObjectInterface {
 
   // TODO BBS-SAL: We run with public access for now as some legacy-code still
   // needs directly access, but should be changed to protected when SAL is done.
-  public $id;
-  public $ownerId;
-  public $sourceId;
-  public $title;
-  public $shortTitle;
+  public $abstract;
```
Comment (kasperg) [inline modules/ting/src/TingObject.php:18]: How about removing the comment above here?

We will not change the visibility of the properties due to all the automagic happening.

File: modules/ting/src/TingObjectInterface.php [hunk 12]
```diff
@@ -318,4 +322,95 @@ public function getType();
...
+  public function getAbstract();
+
+  /**
+   * Returns the creators of the material.
+   *
+   * For a book the creator would typically be the author.
+   *
+   * @param string $format
+   *   TingObjectInterface::NAME_FORMAT_* formats to specify how the authors
+   *   names should be formatted.s
```
Comment (kasperg) [inline modules/ting/src/TingObjectInterface.php]: Trailing s here.

File: modules/ting/src/TingObjectInterface.php [hunk 13]
```diff
@@ -318,4 +322,95 @@ public function getType();
...
+   *
+   * @return string|FALSE
+   *   The abstract.
+   */
+  public function getAbstract();
+
+  /**
+   * Returns the creators of the material.
+   *
+   * For a book the creator would typically be the author.
```
Comment (kasperg) [inline modules/ting/src/TingObjectInterface.php:345]: :+1:

File: modules/ting/ting.entities.inc [hunk 14]
```diff
@@ -272,110 +240,94 @@ class TingEntity extends DingEntity {
...
+   *   List of subjects, empty if none could be found.
+   */
+  public function getSubjects() {
+    return empty($this->reply) ? [] : $this->reply->getSubjects();
   }
 
-  function getLanguage() {
-    return !empty($this->reply->record['dc:language'][''][0]) ? $this->reply->record['dc:language'][''][0] : FALSE;
+  /**
+   * The language of the material.
```
Comment (kasperg) [inline modules/ting/ting.entities.inc:257]: Perhaps it is worth noting here or elsewhere that the language should be human-readable and not a language code.

File: modules/ting/ting.entities.inc [hunk 15]
```diff
@@ -272,110 +240,94 @@ class TingEntity extends DingEntity {
...
+   * @return string|FALSE
+   *   The source of the material.
+   */
+  public function getAc_source() {
+    return empty($this->reply) ? FALSE : $this->reply->getMaterialSource();
   }
 
   /**
-   * Get ISBN numbers of the object.
+   * Get an description of the material.
```
Comment (kasperg) [inline modules/ting/ting.entities.inc:311]: an => a

File: modules/ting/ting.module [hunk 16]
```diff
@@ -552,7 +552,7 @@ function ting_ding_entity_buttons($type, $entity) {
     $buttons = array();
 
     // TODO BBS-SAL: refactor to getter.
-    if ($entity->online_url) {
+    if ($entity->getOnline_url()) {
```
Comment (kasperg) [inline modules/ting/ting.module]: Remember to remove comment.

File: modules/ting/ting.module [hunk 17]
```diff
@@ -527,7 +527,7 @@ function ting_ding_entity_is($object, $class) {
   if ($class == 'reservable' || $class == 'cartable') {
     $sources = variable_get('ting_reservable_sources', _ting_default_reservable_sources());
     $types = variable_get('ting_reservable_types', _ting_default_reservable_types());
-    if (in_array(drupal_strtolower($object->ac_source), $sources) && in_array(drupal_strtolower($object->type), $types)) {
```
Comment (kasperg) [inline modules/ting/ting.module]: Any reason to change this from magic property to method? It does not seem as if we do so everywhere else.

**Review by kasperg — CHANGES_REQUESTED** [review]
Looks really good. I have some questions and minor corrections here and there.

File: modules/ding_loan/plugins/content_types/loans.inc (hunk 1 above)
Comment (danquah) [reply modules/ding_loan/plugins/content_types/loans.inc]: :+1:

File: modules/ding_loan/plugins/content_types/loans.inc (hunk 2 above)
Comment (danquah) [reply modules/ding_loan/plugins/content_types/loans.inc]: Leaving this to later.

File: modules/opensearch/src/OpenSearchTingObject.php (hunk 3 above)
Comment (danquah) [reply modules/opensearch/src/OpenSearchTingObject.php]: Needed as the magic __get'er don't handle arrays well, will add a comment.

File: modules/opensearch/src/OpenSearchTingObject.php (hunk 4 above)
Comment (danquah) [reply modules/opensearch/src/OpenSearchTingObject.php]: Will refactor ...

File: modules/opensearch/src/OpenSearchTingObject.php (hunk 5 above)
Comment (danquah) [reply modules/opensearch/src/OpenSearchTingObject.php:166]: Agreed :)

File: modules/opensearch/src/OpenSearchTingObject.php (hunk 9 above)
Comment (danquah) [reply modules/opensearch/src/OpenSearchTingObject.php]: :+1:

File: modules/opensearch/src/OpenSearchTingObject.php (hunk 6 above)
Comment (danquah) [reply modules/opensearch/src/OpenSearchTingObject.php]: :+1:

File: modules/opensearch/src/OpenSearchTingObject.php (hunk 10 above)
Comment (danquah) [reply modules/opensearch/src/OpenSearchTingObject.php:434]: getRecordEntry() will also return an array, so we should be good.

File: modules/ting/src/TingObject.php (hunk 11 above)
Comment (danquah) [reply modules/ting/src/TingObject.php:18]: Agreed

File: modules/ting/src/TingObjectInterface.php (hunk 13 above)
Comment (danquah) [reply modules/ting/src/TingObjectInterface.php:345]: I'm learning ;)

File: modules/ting/src/TingObjectInterface.php (hunk 12 above)
Comment (danquah) [reply modules/ting/src/TingObjectInterface.php]: :+1:

File: modules/ting/ting.module (hunk 16 above)
Comment (danquah) [reply modules/ting/ting.module]: :+1:

File: modules/ting/ting.module (hunk 17 above)
Comment (danquah) [reply modules/ting/ting.module]: True - might as well keep it consistent

**Review by kasperg — APPROVED** [review]
